src/merge_operator.rs

```diff
@@ -0,0 +1,366 @@
+///         let increment = u64::from_le_bytes(value.as_ref().try_into().unwrap());
+///         Ok(Bytes::copy_from_slice(&(existing + increment).to_le_bytes()))
+///     }
+/// }
+/// ```
+pub trait MergeOperator {
+    /// Merges the existing value with a new value to produce a combined result.
+    ///
+    /// This method is called during reads and compactions to combine multiple merge operands
+    /// into a single value. The implementation must be associative to ensure correct behavior.
+    ///
+    /// # Arguments
+    /// * `existing_value` - The current accumulated value
+    /// * `value` - The new value to merge with the existing value
+    ///
+    /// # Returns
+    /// * `Ok(Bytes)` - The merged result as bytes
+    /// * `Err(MergeOperatorError)` - If the merge operation fails
+    fn merge(&self, existing_value: Bytes, value: Bytes) -> Result<Bytes, MergeOperatorError>;
+}
+
+pub(crate) type MergeOperatorType = Arc<dyn MergeOperator + Send + Sync>;
+
+pub(crate) struct MergeOperatorIterator<T: KeyValueIterator> {
+    merge_operator: MergeOperatorType,
+    delegate: T,
+    /// Entry from the delegate that we've peeked ahead and buffered.
+    buffered_entry: Option<RowEntry>,
+    /// Whether to merge entries with different expire timestamps.
+    merge_different_expire_ts: bool,
+}
+
+#[allow(unused)]
+impl<T: KeyValueIterator> MergeOperatorIterator<T> {
+    pub(crate) fn new(
+        merge_operator: MergeOperatorType,
+        delegate: T,
+        merge_different_expire_ts: bool,
+    ) -> Self {
+        Self {
+            merge_operator,
+            delegate,
+            buffered_entry: None,
+            merge_different_expire_ts,
+        }
+    }
+}
+
+impl<T: KeyValueIterator> MergeOperatorIterator<T> {
+    async fn merge_with_older_entries(
+        &mut self,
+        first_entry: RowEntry,
+    ) -> Result<Option<RowEntry>, SlateDBError> {
+        let mut merged_value = match first_entry.value {
+            ValueDeletable::Merge(ref v) => v.clone(),
+            _ => unreachable!("Entry doesn't contain merge operand."),
+        };
+        let key = first_entry.key;
+        let mut max_create_ts = first_entry.create_ts;
+        let mut min_expire_ts = first_entry.expire_ts;
+
+        // Keep looking ahead and merging as long as we find mergeable entries
+        loop {
+            let next = self.delegate.next_entry().await?;
+            match next {
+                Some(next_entry)
+                    if key == next_entry.key
+                        && (self.merge_different_expire_ts
+                            || first_entry.expire_ts == next_entry.expire_ts) =>
+                {
+                    // Accumulate timestamps. For create_ts we use the maximum (when the accumulated value has last changed),
+                    // and for expire_ts we use the minimum (when the accumulated becomes invalid).
```

Comment:
So, semantically, if I had:

```
merge_different_expire_ts=true

k1, merge1, expire_ts=123
k1, merge2, expire_ts=120
k1, merge3, expire_ts=124
```

The `expire_ts` would resolve to `120`, and all three merges would be dropped as soon as the time `120` is passed, right? That seems a bit surprising to me vs. expiring `merge2` at `120`, `merge1` at `123`, and so on. I suppose if users wanted that, they would not set `merge_different_expire_ts`, though. But there's still no way for users to say they want to merge for reads, but expire by row, which seems like a reasonable use case.

I feel like we might have discussed this in the RFC, but I'm failing to find it now.

Reply:
See TTL Storage and Expiration section of the RFC.

Reply:
Nope, that explains things. Decoupling read and compaction makes sense. 👍

Reply:
Just adding a note to link this to #322 which will probably need to make sure that expired merge elements should likely be filtered out at this point as well.

I do think it wouldn't hurt to expand this comment with the results of this discussion!